File: app/src/main/java/com/xenris/liquidwarsos/activities/GameClientActivity.java


package com.xenris.liquidwarsos.activities;

import android.app.AlertDialog;
import android.os.Bundle;
import android.view.Window;
import android.view.WindowManager;
import android.view.MotionEvent;
import android.view.KeyEvent;
import android.view.Gravity;
import android.widget.TextView;
import android.widget.Toast;
import android.content.Context;
import android.content.DialogInterface;

import androidx.appcompat.app.AppCompatActivity;

import com.xenris.liquidwarsos.Client;
import com.xenris.liquidwarsos.MyGLSurfaceView;
import com.xenris.liquidwarsos.MyRenderer;
import com.xenris.liquidwarsos.NativeInterface;
import com.xenris.liquidwarsos.R;
import com.xenris.liquidwarsos.StaticBits;
import com.xenris.liquidwarsos.Util;

import java.lang.Thread;
import java.util.ArrayList;

public class GameClientActivity extends AppCompatActivity implements Client.ClientCallbacks, Runnable, MyGLSurfaceView.SurfaceCallbacks {
    private MyGLSurfaceView myGLSurfaceView = null;
    private boolean running;
    private int gameStep = -1;
    private short[] xs = new short[5];
    private short[] ys = new short[5];
    private int touchReduction = 5;
    private Context context;
    private boolean usingNativeStateLock = false;
    private boolean gameFinished;
    private boolean lostGame;
    private AlertDialog dialog;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        context = this;

        requestWindowFeature(Window.FEATURE_NO_TITLE);
        final int fullscreen = WindowManager.LayoutParams.FLAG_FULLSCREEN;
        final int keepOn = WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON;
        getWindow().addFlags(fullscreen | keepOn);

        setContentView(R.layout.game);
        myGLSurfaceView = (MyGLSurfaceView)findViewById(R.id.mySurfaceView);
        myGLSurfaceView.setSurfaceCallbacks(this);

        NativeInterface.init(getAssets());
        NativeInterface.createGame(StaticBits.team, StaticBits.map, StaticBits.seed, StaticBits.dotsPerTeam);

        StaticBits.client.setCallbacks(this);

        new Thread(this).start();
    }

    @Override
    public void onPause() {
        super.onPause();
        if(myGLSurfaceView != null)
            myGLSurfaceView.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
        if(myGLSurfaceView != null)
            myGLSurfaceView.onResume();
    }

    @Override
    public void onDestroy() {
        if(!isFinishing())
            finish();
        super.onDestroy();
        if(StaticBits.client != null)
            StaticBits.client.setCallbacks(StaticBits.clientGameSetupActivity);
        if(myGLSurfaceView != null)
            myGLSurfaceView.onPause();
        running = false;
        if(dialog != null)
            dialog.dismiss();
    }

    @Override
    public void run() {
        running = true;
        gameFinished = false;
        lostGame = false;
        StaticBits.client.send(StaticBits.CLIENT_READY, 0);
        while(running) {
            StaticBits.client.send(StaticBits.CLIENT_CURRENT_GAMESTEP, gameStep);
            checkForWinner();
            checkIfLost();
            try { Thread.sleep(50); } catch (InterruptedException ie) { }
        }
        while(usingNativeStateLock)
            try { Thread.sleep(3); } catch (InterruptedException ie) { }
        NativeInterface.destroyGame();
        NativeInterface.uninit();
        StaticBits.client.send(StaticBits.CLIENT_EXIT, 0);
        finish();
    }

    private void checkIfLost() {
        if(lostGame)
            return;
        if(NativeInterface.teamScore(StaticBits.team) == 0) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    lostGame = true;
                    if(dialog != null)
                        dialog.dismiss();
                    AlertDialog.Builder builder = new AlertDialog.Builder(context);
                    builder.setMessage("You Lose");
                    dialog = builder.show();
                    TextView messageText = (TextView)dialog.findViewById(android.R.id.message);
                    messageText.setGravity(Gravity.CENTER);
                    messageText.setTextColor(Util.teamToColour(StaticBits.team));
                    dialog.setCanceledOnTouchOutside(false);
                    Util.makeDialogCancelableIn(dialog, 1500);
                    Util.makeDialogDismissIn(dialog, 5000);
                }
            });
        }
    }

    private void checkForWinner() {
        if(gameFinished)
            return;
        for(int i = 0; i < 6; i++) {
            if(NativeInterface.teamScore(i) == StaticBits.NUMBER_OF_TEAMS*StaticBits.dotsPerTeam) {
                final int p = i;
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        gameFinished = true;
                        if(dialog != null)
                            dialog.dismiss();
                        AlertDialog.Builder builder = new AlertDialog.Builder(context);
                        if(p == StaticBits.team)
                            builder.setMessage("You Win!");
                        else
                            builder.setMessage(Util.teamToNameString(p) + " Wins");
                        dialog = builder.show();
                        TextView messageText = (TextView)dialog.findViewById(android.R.id.message);
                        messageText.setGravity(Gravity.CENTER);
                        messageText.setTextColor(Util.teamToColour(p));
                        dialog.setCanceledOnTouchOutside(false);
                        Util.makeDialogCancelableIn(dialog, 1500);
                        Util.makeDialogDismissIn(dialog, 5000);
                    }
                });
                break;
            }
        }
    }

    private void sendPlayerData() {
        int[] data = new int[5 + 5 + 1];
        for(int i = 0; i < 5; i++) {
            data[i+1] = xs[i];
            data[i+5+1] = ys[i];
        }
        data[0] = StaticBits.PLAYER_POSITION_DATA;
        StaticBits.client.send(data.length, data);
    }

    @Override
    public void onTouch(MotionEvent event) {
        if(event.getAction() == MotionEvent.ACTION_MOVE)
            if(touchReduction-- != 0)
                return;
            else
                touchReduction = 5;

        final int count = event.getPointerCount();
        for(int i = 0; i < 5; i++) {
            if(i < count) {
                xs[i] = (short)((event.getX(i) / (float) MyRenderer.displayWidth) * (float)MyRenderer.WIDTH);
                ys[i] = (short)((MyRenderer.HEIGHT-1) - ((event.getY(i) / (float)MyRenderer.displayHeight) * (float)MyRenderer.HEIGHT));
            } else {
                xs[i] = -1;
                ys[i] = -1;
            }
        }

        if(event.getAction() == MotionEvent.ACTION_POINTER_UP) { //XXX this appears to be not working.
            final int upIndex = event.getActionIndex();
            final int upId = event.getPointerId(upIndex);
            xs[upId] = -1;
            ys[upId] = -1;
        }
        sendPlayerData();
//        android.util.Log.i("mylog", xs[1] + " " + ys[1]);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if((keyCode == KeyEvent.KEYCODE_BACK) && !gameFinished) {
            DialogInterface.OnClickListener clicker = new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int which) {
                    running = false;
                }
            };
            if(dialog != null)
                dialog.dismiss();
            AlertDialog.Builder builder = new AlertDialog.Builder(context);
            builder.setMessage("Back to menu?");
            builder.setPositiveButton("Yes", clicker);
            builder.setNegativeButton("No", null);
            dialog = builder.show();
            TextView messageText = (TextView)dialog.findViewById(android.R.id.message);
            messageText.setGravity(Gravity.CENTER);
            return true;
        }
        else {
            return super.onKeyDown(keyCode, event);
        }
    }

    @Override
    public void onServerMessageReceived(int argc, int[] args) {
        if(args[0] == StaticBits.STEP_GAME) {
            if(running) {
                if(args[2] == (gameStep + 1)) {
                    loadPlayerPositions(args, 3);
                    stepGame(args[1]);
                } else if(args[2] > (gameStep + 1)) {
                    StaticBits.client.send(StaticBits.RESEND_STEPS, gameStep + 1);
                }
            }
        } else if(args[0] == StaticBits.CLIENT_READY_QUERY) {
            if(running)
                StaticBits.client.send(StaticBits.CLIENT_READY, 0);
        } else if(args[0] == StaticBits.KILL_GAME) {
            ToastOnUiThread("Game canceled");
            running = false;
        } else if(args[0] == StaticBits.BACK_TO_MENU) {
            running = false;
        } else if(args[0] == StaticBits.OUT_OF_TIME) {
            outOfTimeMessage(args[1]);
        } else if(args[0] == StaticBits.TIME_DIFF) {
            NativeInterface.setTimeSidebar((float)args[1]/(float)StaticBits.timeLimit);
        }
    }

    @Override
    public void onServerConnectionMade(int id, String ip) {
    }

    @Override
    public void onServerConnectionFailed(String ip) {
        StaticBits.gameWasDisconnected = true;
        ToastOnUiThread("Connection lost");
        running = false;
    }

    @Override
    public void onServerConnectionClosed(String ip) {
        StaticBits.gameWasDisconnected = true;
        ToastOnUiThread("Connection lost");
        running = false;
    }

    private void ToastOnUiThread(final String s) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(context, s, Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void outOfTimeMessage(final int winningTeam) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                gameFinished = true;
                if(dialog != null)
                    dialog.dismiss();
                AlertDialog.Builder builder = new AlertDialog.Builder(context);
                if(winningTeam == StaticBits.team)
                    builder.setMessage("Out of time! You win!");
                else
                    builder.setMessage("Out of time! " + Util.teamToNameString(winningTeam) + " wins!");
                dialog = builder.show();
                TextView messageText = (TextView)dialog.findViewById(android.R.id.message);
                messageText.setTextColor(Util.teamToColour(winningTeam));
                messageText.setGravity(Gravity.CENTER);
                dialog.setCanceledOnTouchOutside(false);
                Util.makeDialogCancelableIn(dialog, 1500);
                Util.makeDialogDismissIn(dialog, 5000);
            }
        });
    }

    private void loadPlayerPositions(int[] data, int offset) {
        short[] tempxs = new short[5];
        short[] tempys = new short[5];
        for(int p = 0; p < 6; p++) {
            for(int xy = 0; xy < 5; xy++) {
                tempxs[xy] = (short)data[offset++];
                tempys[xy] = (short)data[offset++];
            }
            NativeInterface.setPlayerPosition(p, tempxs, tempys);
        }
    }

    private void stepGame(int speed) {
        if(usingNativeStateLock)
            return;
        usingNativeStateLock = true;
        for(int i = 0; i < 10; i++) {
            long previousTime = System.nanoTime();
            NativeInterface.stepDots();
            if(speed == StaticBits.REGULATED_STEP)
                Util.regulateSpeed(previousTime, StaticBits.GAME_SPEED);
        }
        usingNativeStateLock = false;
        gameStep++;
    }
}

File: app/src/main/java/com/dergoogler/liquidwars/activities/GameClientActivity.kt
package com.dergoogler.liquidwars.activities

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.Gravity
import android.view.KeyEvent
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import android.widget.Toast
import com.dergoogler.liquidwars.Client.ClientCallbacks
import com.dergoogler.liquidwars.MyGLSurfaceView
import com.dergoogler.liquidwars.MyGLSurfaceView.SurfaceCallbacks
import com.dergoogler.liquidwars.MyRenderer
import com.dergoogler.liquidwars.NativeInterface
import com.dergoogler.liquidwars.R
import com.dergoogler.liquidwars.StaticBits
import com.dergoogler.liquidwars.Util.makeDialogCancelableIn
import com.dergoogler.liquidwars.Util.makeDialogDismissIn
import com.dergoogler.liquidwars.Util.regulateSpeed
import com.dergoogler.liquidwars.Util.teamToColour
import com.dergoogler.liquidwars.Util.teamToNameString

class GameClientActivity : LiquidCompatActivity(), ClientCallbacks, Runnable, SurfaceCallbacks {
    private var myGLSurfaceView: MyGLSurfaceView? = null
    private var running = false
    private var gameStep = -1
    private val xs = ShortArray(5)
    private val ys = ShortArray(5)
    private var touchReduction = 5
    private var context: Context? = null
    private var usingNativeStateLock = false
    private var gameFinished = false
    private var lostGame = false
    private var dialog: AlertDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.hideSystemUI()
        this.keepOn()
        context = this


        setContentView(R.layout.game)
        myGLSurfaceView = findViewById(R.id.mySurfaceView)
        myGLSurfaceView?.setSurfaceCallbacks(this)

        NativeInterface.init(assets)
        NativeInterface.createGame(
            StaticBits.team,
            StaticBits.map,
            StaticBits.seed,
            StaticBits.dotsPerTeam
        )

        StaticBits.client!!.setCallbacks(this)

        Thread(this).start()
    }

    override fun onPause() {
        super.onPause()
        if (myGLSurfaceView != null) myGLSurfaceView!!.onPause()
    }

    override fun onResume() {
        super.onResume()
        if (myGLSurfaceView != null) myGLSurfaceView!!.onResume()
    }

    override fun onDestroy() {
        if (!isFinishing) finish()
        super.onDestroy()
        if (StaticBits.client != null) StaticBits.client!!.setCallbacks(StaticBits.clientGameSetupActivity)
        if (myGLSurfaceView != null) myGLSurfaceView!!.onPause()
        running = false
        if (dialog != null) dialog!!.dismiss()
    }

    override fun run() {
        running = true
        gameFinished = false
        lostGame = false
        StaticBits.client!!.send(StaticBits.CLIENT_READY, 0)
        while (running) {
            StaticBits.client!!.send(StaticBits.CLIENT_CURRENT_GAMESTEP, gameStep)
            checkForWinner()
            checkIfLost()
            try {
                Thread.sleep(50)
            } catch (ie: InterruptedException) {
            }
        }
        while (usingNativeStateLock) try {
            Thread.sleep(3)
        } catch (ie: InterruptedException) {
        }
        NativeInterface.destroyGame()
        NativeInterface.uninit()
        StaticBits.client!!.send(StaticBits.CLIENT_EXIT, 0)
        finish()
    }

    private fun checkIfLost() {
        if (lostGame) return
        if (NativeInterface.teamScore(StaticBits.team) == 0) {
            runOnUiThread {
                lostGame = true
                if (dialog != null) dialog!!.dismiss()
                val builder = AlertDialog.Builder(context)
                builder.setMessage("You Lose")
                dialog = builder.show()
                val messageText = dialog?.findViewById<View>(android.R.id.message) as TextView
                messageText.gravity = Gravity.CENTER
                messageText.setTextColor(teamToColour(StaticBits.team))
                dialog?.setCanceledOnTouchOutside(false)
                makeDialogCancelableIn(dialog!!, 1500)
                makeDialogDismissIn(dialog!!, 5000)
            }
        }
    }

    private fun checkForWinner() {
        if (gameFinished) return
        for (i in 0..5) {
            if (NativeInterface.teamScore(i) == StaticBits.NUMBER_OF_TEAMS * StaticBits.dotsPerTeam) {
                val p = i
                runOnUiThread {
                    gameFinished = true
                    if (dialog != null) dialog!!.dismiss()
                    val builder = AlertDialog.Builder(context)
                    if (p == StaticBits.team) builder.setMessage("You Win!")
                    else builder.setMessage(teamToNameString(p) + " Wins")
                    dialog = builder.show()
                    val messageText = dialog?.findViewById<View>(android.R.id.message) as TextView
                    messageText.gravity = Gravity.CENTER
                    messageText.setTextColor(teamToColour(p))
                    dialog?.setCanceledOnTouchOutside(false)
                    makeDialogCancelableIn(dialog!!, 1500)
                    makeDialogDismissIn(dialog!!, 5000)
                }
                break
            }
        }
    }

    private fun sendPlayerData() {
        val data = IntArray(5 + 5 + 1)
        for (i in 0..4) {
            data[i + 1] = xs[i].toInt()
            data[i + 5 + 1] = ys[i].toInt()
        }
        data[0] = StaticBits.PLAYER_POSITION_DATA
        StaticBits.client!!.send(data.size, data)
    }

    override fun onTouch(event: MotionEvent) {
        if (event.action == MotionEvent.ACTION_MOVE) if (touchReduction-- != 0) return
        else touchReduction = 5

        val count = event.pointerCount
        for (i in 0..4) {
            if (i < count) {
                xs[i] =
                    ((event.getX(i) / MyRenderer.displayWidth.toFloat()) * MyRenderer.WIDTH.toFloat()).toInt()
                        .toShort()
                ys[i] =
                    ((MyRenderer.HEIGHT - 1) - ((event.getY(i) / MyRenderer.displayHeight.toFloat()) * MyRenderer.HEIGHT.toFloat())).toInt()
                        .toShort()
            } else {
                xs[i] = -1
                ys[i] = -1
            }
        }

        if (event.action == MotionEvent.ACTION_POINTER_UP) { //XXX this appears to be not working.
            val upIndex = event.actionIndex
            val upId = event.getPointerId(upIndex)
            xs[upId] = -1
            ys[upId] = -1
        }
        sendPlayerData()
        //        android.util.Log.i("mylog", xs[1] + " " + ys[1]);
    }

    override fun onHover(v: View, event: MotionEvent) {
        onTouch(event)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if ((keyCode == KeyEvent.KEYCODE_BACK) && !gameFinished) {
            val clicker = DialogInterface.OnClickListener { dialog, which -> running = false }
            if (dialog != null) dialog!!.dismiss()
            val builder = AlertDialog.Builder(context)
            builder.setMessage("Back to menu?")
            builder.setPositiveButton("Yes", clicker)
            builder.setNegativeButton("No", null)
            dialog = builder.show()
            val messageText = dialog?.findViewById<View>(android.R.id.message) as TextView
            messageText.gravity = Gravity.CENTER
            return true
        } else {
            return super.onKeyDown(keyCode, event)
        }
    }

    override fun onServerMessageReceived(argc: Int, args: IntArray) {
        if (args[0] == StaticBits.STEP_GAME) {
            if (running) {
                if (args[2] == (gameStep + 1)) {
                    loadPlayerPositions(args, 3)
                    stepGame(args[1])
                } else if (args[2] > (gameStep + 1)) {
                    StaticBits.client!!.send(StaticBits.RESEND_STEPS, gameStep + 1)
                }
            }
        } else if (args[0] == StaticBits.CLIENT_READY_QUERY) {
            if (running) StaticBits.client!!.send(StaticBits.CLIENT_READY, 0)
        } else if (args[0] == StaticBits.KILL_GAME) {
            ToastOnUiThread("Game canceled")
            running = false
        } else if (args[0] == StaticBits.BACK_TO_MENU) {
            running = false
        } else if (args[0] == StaticBits.OUT_OF_TIME) {
            outOfTimeMessage(args[1])
        } else if (args[0] == StaticBits.TIME_DIFF) {
            NativeInterface.setTimeSidebar(args[1].toFloat() / StaticBits.timeLimit.toFloat())
        }
    }

    override fun onServerConnectionMade(id: Int, ip: String) {
    }

    override fun onServerConnectionFailed(ip: String) {
        StaticBits.gameWasDisconnected = true
        ToastOnUiThread("Connection lost")
        running = false
    }

    override fun onServerConnectionClosed(ip: String) {
        StaticBits.gameWasDisconnected = true
        ToastOnUiThread("Connection lost")
        running = false
    }

    private fun ToastOnUiThread(s: String) {
        runOnUiThread { Toast.makeText(context, s, Toast.LENGTH_SHORT).show() }
    }

    private fun outOfTimeMessage(winningTeam: Int) {
        runOnUiThread {
            gameFinished = true
            if (dialog != null) dialog!!.dismiss()
            val builder = AlertDialog.Builder(context)
            if (winningTeam == StaticBits.team) builder.setMessage("Out of time! You win!")
            else builder.setMessage("Out of time! " + teamToNameString(winningTeam) + " wins!")
            dialog = builder.show()
            val messageText = dialog?.findViewById<View>(android.R.id.message) as TextView
            messageText.setTextColor(teamToColour(winningTeam))
            messageText.gravity = Gravity.CENTER
            dialog?.setCanceledOnTouchOutside(false)
            makeDialogCancelableIn(dialog!!, 1500)
            makeDialogDismissIn(dialog!!, 5000)
        }
    }

    private fun loadPlayerPositions(data: IntArray, offset: Int) {
        var offset = offset
        val tempxs = ShortArray(5)
        val tempys = ShortArray(5)
        for (p in 0..5) {
            for (xy in 0..4) {
                tempxs[xy] = data[offset++].toShort()
                tempys[xy] = data[offset++].toShort()
            }
            NativeInterface.setPlayerPosition(p, tempxs, tempys)
        }
    }

    private fun stepGame(speed: Int) {
        if (usingNativeStateLock) return
        usingNativeStateLock = true
        for (i in 0..9) {
            val previousTime = System.nanoTime()
            NativeInterface.stepDots()
            if (speed == StaticBits.REGULATED_STEP) regulateSpeed(
                previousTime,
                StaticBits.GAME_SPEED
            )
        }
        usingNativeStateLock = false
        gameStep++
    }
}
